terraform/eval_for_each.go

```diff
@@ -0,0 +1,64 @@
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/terraform/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// evaluateResourceForEachExpression interprets a "for_each" argument on a resource.
+//
+// Returns a cty.Value map, and diagnostics if necessary
+func evaluateResourceForEachExpression(expr hcl.Expression, ctx EvalContext) (forEach map[string]cty.Value, diags tfdiags.Diagnostics) {
+	if expr == nil {
+		return nil, nil
```

Comment:
It looks like this is using `nil` to signal that `for_each` isn't set, vs. an empty map to signal that `for_each` is set to an expression that returned an empty map.

That does seem okay to me, but the usual idiom/convention in Go is for a `nil` map and an empty map to be considered equivalent, so I think worth noting that specifically in the doc comment so that the reader can see that we're making a distinction that wouldn't normally be significant.

Reply:
I've noted in the doc comment how this using nil -- since `count` evaluation is using `-1` as its "special value" I believe it makes sense to do that here as well.

Reply:
LGTM!